aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -208,6 +340,539 @@ void cpu_upsample_nearest_channels_last(
+
+          for (int j=0; j<interp_size; j++) {
+            idx_ptr = output[2 * j + 0].data_ptr<int64_t>();
+            idx_ptr[i] = static_cast<int64_t>(std::max(std::min(input_index + j - 1, input_size - 1), zero)) * stride;
+            wt_ptr = output[2 * j + 1].data_ptr<scalar_t>();
+            wt_ptr[i] = coeffs[j];
+          }
+        }
+      }
+    );
+    return output;
+  }
+};
+
+// Generic upsampling interpolation kernel for N-d case.
+// Input is assumed to be like NCHW, NCL, NCKHW - interpolated spatial dimension
+// are those from the end up to batch size N and number of channels C.
+//
+// Internally, it uses TensorIterator to optimize the computations.
+// - out_ndims is the number of interpolated dims: 1, 2, 3
+// - scale_type is template type for scales, typically c10::optional<double>
+// - template<typename> class F is one of the above structs to compute indices and weights
+template <int out_ndims, typename scale_type, class F>
+void upsample_generic_Nd_kernel_impl(
+    const Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    const scale_type& scales) {
+
+  // input can be NCHW, NCL or NCKHW
+  auto shape = input.sizes().vec();
+  auto strides = input.strides().vec();
+  auto oshape = output.sizes();
+
+  TORCH_INTERNAL_ASSERT(
+    shape.size() == oshape.size() && shape.size() == 2 + out_ndims
+  );
+  TORCH_INTERNAL_ASSERT(strides.size() == 2 + out_ndims);
+
+  for (int i=0; i<out_ndims; i++) {
+    shape[i + 2] = oshape[i + 2];
+    strides[i + 2] = 0;
+  }
+  auto restrided_input = input.as_strided(shape, strides);
+
+  std::vector<std::vector<Tensor>> indices_weights;
+
+  constexpr int interp_size = F::interp_size;
+  auto input_scalar_type = input.scalar_type();
+  if (interp_size == 1 && input_scalar_type == at::ScalarType::Byte) {
+    // nearest also supports uint8 tensor, but we have to use float
+    // with compute_indices_weights
+    input_scalar_type = at::ScalarType::Float;
+  }
+
+  for (int i=0; i<out_ndims; i++) {
+    indices_weights.emplace_back(
+      F::compute_indices_weights(
+        input_scalar_type, input.size(i + 2), oshape[i + 2],
+        input.stride(i + 2) * input.element_size(),
+        input.dim(), i + 2, align_corners, scales[i]
+      )
+    );
+  }
+
+  TensorIteratorConfig config;
+  config.check_all_same_dtype(false)
+    .declare_static_dtype_and_device(input.scalar_type(), input.device())
+    .add_output(output)
+    .add_input(restrided_input);
+
+  for (auto & idx_weight: indices_weights) {
+    for (auto& tensor : idx_weight) {
+      config.add_input(tensor);
+    }
+  }
+
+  auto iter = config.build();
+
+  if (interp_size > 1) {
+    // Nearest also supports uint8 tensor, so need to handle it separately
+    AT_DISPATCH_FLOATING_TYPES(
+        iter.dtype(), "upsample_generic_Nd", [&] {
+        // MSVC can not catch constexpr int interp_size here
+        constexpr int mode = F::interp_size;
```

Comment:
looks like this is duplicate from  `constexpr int interp_size = F::interp_size;` ?

Reply:
This is due to MSVC compiler. See 
- https://github.com/pytorch/pytorch/pull/54500/commits/d622a998fef46581b1c0cf0955ad61bccaa8cafd
- https://app.circleci.com/pipelines/github/pytorch/pytorch/289645/workflows/ec4a7acd-7631-4964-8d8e-ed513bba63db/jobs/11783205